### DISCUSSION TITLE
SQLSTATE[HY000] [1045] Access denied for user

### DISCUSSION THREAD

**Discussion by dynamicat** [firefly-iii/firefly-iii#6875] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [ ] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I have passed the .env file manually and then also passed the env variables like shown in the compose file.  The logs show that the DB is available but says it needs to wait for the DB to boot.  The database is on a docker container also and is used by other services so I know it is running and functions.

### Expected behaviour

`
version: '3.8'

services:
  firefly:
    image: fireflyiii/core:latest
    restart: always
    volumes:
      - firefly_upload:/var/www/html/storage/upload
    networks:
      - dbnet
      - t2_proxy
    environment:
      - APP_KEY=appkey
      - DEFAULT_LANGUAGE=en_US
      - TZ=Europe/Prague
      - DB_CONNECTION=mysql
      - DB_HOST=mariadb
      - DB_PORT=3306
      - DB_DATABASE=firefly
      - DB_USERNAME=firefly
      - DB_PASSWORD=firefly_db_password
      - APP_URL=https://firefly.mydomain.com
    secrets:
     - firefly_db_password
    deploy:
      mode: global
      placement: 
        constraints: [node.role == manager]     
      labels:
        - "traefik.enable=true"
        ## HTTP Routers
        - "traefik.http.routers.firefly-rtr.entrypoints=https"
        - "traefik.http.routers.firefly-rtr.rule=Host(`firefly.mydomain.com`)"
        - "traefik.http.routers.firefly-rtr.tls=true"
        ## Middlewares
        - "traefik.http.routers.firefly-rtr.middlewares=chain-no-auth@file"
        ## HTTP Services
        - traefik.docker.network=t2_proxy
        - "traefik.http.routers.firefly-rtr.service=firefly-svc"
        - "traefik.http.services.firefly-svc.loadbalancer.server.port=8080"
        - "traefik.http.routers.firefly-rtr.middlewares=chain-authelia@file"

secrets:
  firefly_db_password:
    external: true         

volumes:
   firefly_upload:
    driver: local
    driver_opts:
      o: bind
      device: /mnt/data/app-config/firefly/upload
      type: none

networks:
  dbnet:
    external: true
  t2_proxy:
    external: true
`

### Debug information

`
[previous exception] [object] (PDOException(code: 1045): SQLSTATE[HY000] [1045] Access denied for user 'firefly'@'10.0.7.6' (using password: YES) at /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php:70)
      
[stacktrace]
#0 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php(70): PDO->__construct()
#1 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php(46): Illuminate\\Database\\Connectors\\Connector->createPdoConnection()
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connectors/MySqlConnector.php(24): Illuminate\\Database\\Connectors\\Connector->createConnection()
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connectors/ConnectionFactory.php(184): Illuminate\\Database\\Connectors\\MySqlConnector->connect()
#4 [internal function]: Illuminate\\Database\\Connectors\\ConnectionFactory->Illuminate\\Database\\Connectors\\{closure}()
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php(1181): call_user_func()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php(539): Illuminate\\Database\\Connection->getPdo()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php(753): Illuminate\\Database\\Connection->Illuminate\\Database\\{closure}()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php(720): Illuminate\\Database\\Connection->runQueryCallback()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php(546): Illuminate\\Database\\Connection->run()
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Connection.php(498): Illuminate\\Database\\Connection->statement()
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php(32): Illuminate\\Database\\Connection->insert()
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(3270): Illuminate\\Database\\Query\\Processors\\Processor->processInsertGetId()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1861): Illuminate\\Database\\Query\\Builder->insertGetId()
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1322): Illuminate\\Database\\Eloquent\\Builder->__call()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1287): Illuminate\\Database\\Eloquent\\Model->insertAndSetId()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1128): Illuminate\\Database\\Eloquent\\Model->performInsert()
#17 /var/www/html/vendor/laravel/passport/src/ClientRepository.php(153): Illuminate\\Database\\Eloquent\\Model->save()
#18 /var/www/html/vendor/laravel/passport/src/ClientRepository.php(168): Laravel\\Passport\\ClientRepository->create()
#19 /var/www/html/vendor/laravel/passport/src/Console/ClientCommand.php(67): Laravel\\Passport\\ClientRepository->createPersonalAccessClient()
#20 /var/www/html/vendor/laravel/passport/src/Console/ClientCommand.php(43): Laravel\\Passport\\Console\\ClientCommand->createPersonalClient()
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Laravel\\Passport\\Console\\ClientCommand->handle()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\\Container\\BoundMethod::call()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\\Container\\Container->call()
#27 /var/www/html/vendor/symfony/console/Command/Command.php(291): Illuminate\\Console\\Command->execute()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(152): Symfony\\Component\\Console\\Command\\Command->run()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Concerns/CallsCommands.php(68): Illuminate\\Console\\Command->run()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Concerns/CallsCommands.php(28): Illuminate\\Console\\Command->runCommand()
#31 /var/www/html/vendor/laravel/passport/src/Console/InstallCommand.php(42): Illuminate\\Console\\Command->call()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Laravel\\Passport\\Console\\InstallCommand->handle()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\\Container\\BoundMethod::call()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\\Container\\Container->call()
#38 /var/www/html/vendor/symfony/console/Command/Command.php(291): Illuminate\\Console\\Command->execute()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(152): Symfony\\Component\\Console\\Command\\Command->run()
#40 /var/www/html/vendor/symfony/console/Application.php(1014): Illuminate\\Console\\Command->run()
#41 /var/www/html/vendor/symfony/console/Application.php(301): Symfony\\Component\\Console\\Application->doRunCommand()
#42 /var/www/html/vendor/symfony/console/Application.php(171): Symfony\\Component\\Console\\Application->doRun()
#43 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Application.php(102): Symfony\\Component\\Console\\Application->run()
#44 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(155): Illuminate\\Console\\Application->run()
#45 /var/www/html/artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle()
#46 {main}
"} 
   Illuminate\Database\QueryException 
  SQLSTATE[HY000] [1045] Access denied for user 'firefly'@'10.0.7.6' (using password: YES) (SQL: insert into `oauth_clients` (`user_id`, `name`, `secret`, `provider`, `redirect`, `personal_access_client`, `password_client`, `revoked`, `updated_at`, `created_at`) values (?, Firefly III Personal Access Client, PIye69FQGAmQ7PICbcfRpDFgzQQHSqWXVKg2B96Q, ?, http://localhost, 1, 0, 0, 2023-01-16 08:11:34, 2023-01-16 08:11:34))
  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
    756▕         // If an exception occurs when attempting to run a query, we'll format the error
    757▕         // message to include the bindings with SQL, which will make this exception a
    758▕         // lot more helpful to the developer instead of just the database's errors.
    759▕         catch (Exception $e) {
  ➜ 760▕             throw new QueryException(
    761▕                 $query, $this->prepareBindings($bindings), $e
    762▕             );
    763▕         }
    764▕     }
      +46 vendor frames 
  47  artisan:37
      Illuminate\Foundation\Console\Kernel::handle()
Updated version.
[2023-01-16 08:11:36] local.INFO: Could not create or generate installation ID. Do not continue.  
`

**Comment by JC5** [comment, marked as answer]
It looks like you submit a secret to the docker container but don't use it in the environment variables, is that correct?

**Reply by dynamicat** [reply]
Yes, is it okay to do it like that?

**Reply by JC5** [reply]
Of course but you need to refer to it, otherwise it won't be picked up. I'm not sure where Docker secrets are stored exactly in the container but try this instead of DB_PASSWORD:

```
-  DB_PASSWORD_FILE=/var/secrets/firefly_db_password
```

**Reply by dynamicat** [reply]
I did that but I get the same error:

`
      - DB_CONNECTION=mysql
      - DB_HOST=mariadb
      - DB_PORT=3306
      - DB_DATABASE=firefly
      - DB_USERNAME=firefly
      - DB_PASSWORD_FILE=/var/secrets/firefly_db_password
   #   - DB_PASSWORD=firefly_db_password
`

**Reply by dynamicat** [reply]
It works, I changed it to:

`
- DB_PASSWORD_FILE=/run/secrets/firefly_db_password
`

Add run instead of var
